### ISSUE TITLE
Exception should not be eaten in ApplicationJsonMessageMarshallingConverter?

### ISSUE THREAD

**Issue by Stexxen** [spring-cloud/spring-cloud-stream#2129]
This [commit](https://github.com/spring-cloud/spring-cloud-stream/commit/cedacc8b17e6121342f4945b8c9b3593dbaa550e#diff-57cb16a0b23f07a5289191280301c5b97b6065076656baf21c2aa5e6f35bfcf2R117) causes the exception raised by the converter to be eaten/ignored.

This is for the `ApplicationJsonMessageMarshallingConverter` in package `org.springframework.cloud.stream.converter`

I'm not sure what issue this was put in place to fix but now it no longer raises exceptions with application/json messages that contain invalid JSON

Here is a test case
```java
package org.springframework.cloud.stream.converter;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.springframework.messaging.support.GenericMessage;

import static org.junit.jupiter.api.Assertions.assertNotNull;

public class SampleTest {


  @Test
  void badJson() {
    ApplicationJsonMessageMarshallingConverter converter = new ApplicationJsonMessageMarshallingConverter(initObjectMapper());
    Object thing = converter.convertFromInternal(new GenericMessage<>("{ notjson }".getBytes()), JsonNode.class, null);
    assertNotNull(thing);
  }

  private ObjectMapper initObjectMapper() {
    ObjectMapper objectMapper = new ObjectMapper();
    objectMapper.configure(MapperFeature.DEFAULT_VIEW_INCLUSION, false);
    objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    return objectMapper;
  }

}
```

The initObjectMapper() method is just [copied directly from](https://github.com/spring-projects/spring-framework/blob/master/spring-messaging/src/main/java/org/springframework/messaging/converter/MappingJackson2MessageConverter.java#L91-L96) `MappingJackson2MessageConverter` to replicate how ApplicationJsonMessageMarshallingConverter is initialised to create this unit test.

Jackson raises a standard `JsonParseException`, which is wrapped in a `MessageConversionException` in MappingJackson2MessageConverter which is then ignored.

In normal operations because no exception is raised and convertFromInternal now returns a `null`, execution continues and eventually fails elsewhere with an exception like the following
```
Caused by: java.lang.IllegalArgumentException: Failed to convert input: GenericMessage [payload=byte[269], headers={kafka_offset=23200, kafka_consumer=org.apache.kafka.clients.consumer.KafkaConsumer@58d8b818, deliveryAttempt=3, kafka_timestampType=CREATE_TIME, kafka_receivedPartitionId=0, kafka_receivedTopic=tibgenericevents, kafka_receivedTimestamp=1614248104312, contentType=application/json, kafka_groupId=original-event4}] to class [classname]
	at org.springframework.util.Assert.notNull(Assert.java:201) ~[spring-core-5.3.4.jar:5.3.4]
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.convertInputIfNecessary(SimpleFunctionRegistry.java:881) ~[spring-cloud-function-context-3.1.0.jar:3.1.0]
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.doApply(SimpleFunctionRegistry.java:573) ~[spring-cloud-function-context-3.1.0.jar:3.1.0]
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.apply(SimpleFunctionRegistry.java:428) ~[spring-cloud-function-context-3.1.0.jar:3.1.0]
	at org.springframework.cloud.stream.function.PartitionAwareFunctionWrapper.apply(PartitionAwareFunctionWrapper.java:79) ~[spring-cloud-stream-3.1.0.jar:3.1.0]
	at org.springframework.cloud.stream.function.FunctionConfiguration$FunctionWrapper.apply(FunctionConfiguration.java:713) ~[spring-cloud-stream-3.1.0.jar:3.1.0]
	at org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1.handleMessageInternal(FunctionConfiguration.java:559) ~[spring-cloud-stream-3.1.0.jar:3.1.0]
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:56) ~[spring-integration-core-5.4.4.jar:5.4.4]
```

Although it could be said that essence of the exception "failure to convert" is correct 
The 'why' is lost and much time is spent then trying to figure out why an assert.NotNull exception is raised in `SimpleFunctionRegistry` 

In summary, if Jackson raises *any* exception during the conversation of the payload it is ignored, which I don't think is correct.

I should also mention that the version of `ApplicationJsonMessageMarshallingConverter` in `org.springframework.cloud.function.context.config` does not contain this change, and the exception is not ignored.

**Comment by rafaelbrizola** [comment]
I second this. In my case it's a Kinesis consumer and the error message was even more cryptic:

```
Caused by: java.lang.ClassCastException: class [B cannot be cast to class com.Myclass([B is in module java.base of loader 'bootstrap'; com.Myclass is in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @7494bd6a)
```

**Comment by dharezlak** [comment]
I am experiencing something similar in case of the Kafka binder and cloud dependencies `2020.0.3`. If I define a consumer like this:
```
@Bean
public Consumer<Message<Dto>> consumer() {
  return message -> {
    System.out.println("Got: " + message.getPayload());
  };
}
```
... and send in something that should not deserialize properly my payload will be output as a byte array (e.g. `Got: [B@2d35a0e8` which also means that the handler method will get invoked when it should not). But if I do something like this (no more `Message` type wrapper):
```
@Bean
public Consumer<Dto> consumer() {
  return message -> {
    System.out.println("Got: " + message);
  };
}
```
... the handler method will not be invoked and SCS will throw an exception and, if configured, the DLQ will be used.

Additionally, if I do something like this with the message in the first case: 
```
message.getPayload().getClass().getName()
```
... I will get the following exception:
```
Caused by: java.lang.ClassCastException: class [B cannot be cast to class dto.Dto ([B is in module java.base of loader 'bootstrap'; dto.Dto is in unnamed module of loader 'app')
```

**Comment by rafaelbrizola** [comment]
My workaround was to consume the message as String and do the unmarshalling manually later. (Our kinesis stream has json in it) At least, this way, we could log the message and figure out what was wrong with it.

**Comment by Stexxen** [comment]
@olegz Thanks for the fix.

Can I suggest the small change to the test here - https://github.com/spring-cloud/spring-cloud-stream/commit/262c088b943324a0f4f9d431dfb22a023f7da5cf#commitcomment-52804831
